Create async search index if necessary on updates and deletes

jimczi opened the pull request:
This change ensures that we create the async search index with the right mappings and settings when updating or deleting a document. Users can delete the async search index at any time so we have to re-create it internally if necessary before applying any new operation.

jimczi committed:
Register a template for .async-search index

This change registers a template for the .async-search system index in order
to ensure that deleting and re-creating the index outside of the async search API
doesn't break the strict mapping.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

jimczi committed:
fix test compilation

jimczi committed:
fix template name in ESRestTestCase

jimczi committed:
line len

imotov commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/AsyncSearchUser.java:
```diff
@@ -16,7 +16,8 @@
     public static final Role ROLE = Role.builder(new RoleDescriptor(ROLE_NAME,
-            null,
+            // handles template registry for the .async-search index
+            new String[] { "manage_index_templates" },
```
Hmm, does this meant that any async search user can now mess with templates. Does this really necessary?

imotov reviewed (commented):
LGTM in general, but I wonder if we can somehow avoid extending async search role with index template privilege.

jimczi committed:
Merge branch 'master' into async_search_index_template

jimczi committed:
apply review comment and set index.codec to best_compression

imotov commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskTemplateRegistry.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    protected String getOrigin() {
+        return ClientHelper.INDEX_LIFECYCLE_ORIGIN;
```
It looks like we are reusing this in snapshot as well. So I am ok with this approach. I am just not 100% sure of it needs a separate original and a separate role. Maybe ping somebody from security?

jimczi committed:
another round of review

jimczi committed:
remove change

jimczi commented:
@imotov , we discussed with @albertzaharovits and agreed that an index template is not needed. Instead, we ensure that the index is re-created when necessary on every operation (index, delete and update). We do the same for the security index so that's consistent. I updated the title, description and modified the PR to reflect the last discussions. Can you take another look ?

jimczi committed:
add a test case

imotov commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java:
```diff
@@ -225,7 +227,7 @@ public void deleteResponse(AsyncExecutionId asyncExecutionId,
         try {
             DeleteRequest request = new DeleteRequest(index).id(asyncExecutionId.getDocId());
-            client.delete(request, listener);
+            createIndexIfNecessary(ActionListener.wrap(v -> client.delete(request, listener), listener::onFailure));
```
Do we need to create an index in this case or this is for consistency?

jimczi replied:
I pushed https://github.com/elastic/elasticsearch/pull/64606/commits/a273d920cd9f91e67507abe2c1c321c3e9eb62ad to explain the reasoning.

imotov replied:
:+1:

imotov reviewed (approved):
I think reasoning for creating an index while deleting response is not very intuitive. Perhaps, a comment explaining a rationale for it could help.

jimczi committed:
Merge branch 'master' into async_search_index_template

jimczi committed:
apply review comment

jimczi committed:
Merge branch 'master' into async_search_index_template
